Meta: Clearly document current guidance for changing existing PEPs (#2378)
python/peps: pep-0001.txt

PEP Maintenance
---------------

In general, PEPs are no longer substantially modified after they have reached

> the Final, Rejected or Superseded state. Once a PEP has been completed,

Reviewer: This is also true (with only minor exceptions) for the Accepted state.

Author: Yeah; that was my impression and what the original wording of my Contributing Guide addition said, but when I moved it to PEP 1, the present wording there only described this for the "Final" state, and mentioned changes to "Accepted" PEPs on equal footing with "Provisional" ones. I've revised this and the passage below accordingly to reflect the current reality here.

PEP Maintenance
---------------

In general, PEPs are no longer substantially modified after they have reached
the Accepted, Final, Rejected or Superseded state. Once resolution is reached,
a PEP is considered a historical document rather than a living specification.
Formal documentation of the expected behavior should be maintained elsewhere,
such as the `Language Reference`_ for core features, the `Library Reference`_
for standard library modules or the `PyPA Specifications`_ for packaging.

If changes based on implementation experience and user feedback are made to

> Standards track PEPs while in the Provisional or (with SC approval) Accepted

Reviewer 2: Do we specify anywhere else that changes to Accepted PEPs require re-approval? If so, the `(with SC approval)` can be taken out. If not, please can this be rephrased to not have the parenthetical aside.

Author: Not that I could find, but its possible I missed it—this PEP is rather long and its organization sometimes results in things being different places. I was basing this off @gvanrossum 's reply here and what seems to be accepted (heh) practice.

We can do this, if you prefer:

```suggestion
Standards track PEPs while in the Provisional (or with SC approval, Accepted)
```

Off the top of my head, I'm not sure I can think of another straightforward way of avoiding this while still being explicit and unambitious that it only applies to `Accepted` and not `Provisional`, at least without rewriting this whole otherwise unchanged paragraph. Could you maybe suggest something, and/or explain the issue with the parenthetical?

Reviewer 2: > explain the issue with the parenthetical

A general aversion to parenthetical asides in formal writing. Suggestion below, though it is a re-write of the entire paragraph.

A

```rest
While implementing a Standards Track PEP, it may be desirable to alter the
behaviour described in the PEP due to user feedback or implementation 
experience. The PEP should be updated such that it accurately describes
the implementation at the point where it is marked Final. Changes to the
behaviour described in an Accepted PEP must also recieve approval from the 
`Steering Council`_.
```

Author: I'm not sure we need to completely re-write this paragraph from the original just to suit one person's personal aversion to parenthetical asides, but what do others think (@JelleZijlstra @hugovk )?

Reviewer 3: I think it's fine as is.

state, they should be noted in the PEP, such that the PEP accurately describes
the implementation at the point where it is marked Final.

Active Informational and Process PEPs may be updated over time to reflect
changes to development practices and other details. The precise process
followed in these cases will depend on the nature and purpose of the PEP
being updated. Substantive changes to governance PEPs must be reviewed by the

> Steering Council (by opening a `Steering Council issue`_).

Reviewer 2: Again, is this captured elsewhere to your satisfaction? If so the sentence starting "Substantive" can be removed.

Author: I did some searching (before and now) and I couldn't find anywhere else that explicitly mentioned this. With how long PEP 1 is, it is not impossible that I missed it, but at least it didn't appear to be anywhere I could think to find it, and this seems to be the most logical place to put it.

We could also modify this to say

```suggestion
being updated. Substantive changes to Active Process PEPs must be
reviewed by the Steering Council (by opening a `Steering Council issue`_).
```

since that more clearly defines what PEPs must have this approval, since "governance PEPs" are not so well defined.

To note, as I've noticed previously, PEP 13 is inexplicably listed as an Informational PEP rather than a Process PEP, despite it being the Meta/Process PEP of all Process PEPs as it defines the "process" by which the Python community governs itself and in turn decides on the PEP process, and is thus the most immutable of them all.

Reviewer 2: Your suggested wording looks good (although, I would actually interpret "governance PEPs" as just 13 and the 8000 series, so perhaps it would be useful to check what the Council care about enough to add to their workload!

A

Author: Any thoughts on this @brettcannon ?

Reviewer 4: The SC *cannot* unilaterally change PEP 13. The procedure to change PEP 13 is described [in the PEP itself](https://peps.python.org/pep-0013/#changing-this-document). The same goes for the [PyPA governance PEP](https://peps.python.org/pep-0609/#updates-to-the-governance-specification-processes) for example.
IMO the previous wording is enough here, no addition about the SC is necessary.

Author: Right, but this is about Active Process PEPs in general, not anything specific to one PEP, which is what my suggestion above changes it to be more clear on (and, inexplicably, PEP 13 is listed as a mere _"Informational"_ PEP rather than a "Process" PEP, so technically it wouldn't apply to PEP 13 at all, albeit almost certainly due to an oversight). That said, the previous sentence says "The precise process followed in these cases will depend on the nature and purpose of the PEP being updated.", so if there's no need to explicitly specify this here, I'll just remove it.

```suggestion
in question.
```

To note, I originally included another clause specifying that beyond SC review, any guidance in the PEP in question [e.g. PEP 13, 609, etc] also needed to be followed when updating it, but I had elided it in the current version for brevity, due to concerns about the original proposed wording being too verbose and overly specific.

PEP Maintenance
---------------

In general, PEPs are no longer substantially modified after they have reached
the Accepted, Final, Rejected or Superseded state. Once resolution is reached,
a PEP is considered a historical document rather than a living specification.
Formal documentation of the expected behavior should be maintained elsewhere,
such as the `Language Reference`_ for core features, the `Library Reference`_
for standard library modules or the `PyPA Specifications`_ for packaging.

If changes based on implementation experience and user feedback are made to
Standards track PEPs while in the Provisional or (with SC approval) Accepted
state, they should be noted in the PEP, such that the PEP accurately describes
the implementation at the point where it is marked Final.

> Active Informational and Process PEPs may be updated over time to reflect

Reviewer 5: ```suggestion
Active, Informational and Process PEPs may be updated over time to reflect
```

Author: Actually, the intended syntax is `Active (Informational and Process) PEPs`, not `(Active, Informational and Process) PEPs` as the above change would imply. Any suggestions to make that clearer (as that obviously wasn't what came across)? E.g.

```suggestion
Active-status Informational and Process PEPs may be updated over time to reflect
```

or

```suggestion
Informational and Process PEPs that are Active may be updated over time to reflect
```

or even as a literal parenthetical?

```suggestion
Active (Informational and Process) PEPs may be updated over time to reflect
```

Reviewer 3: The last one looks clearest to me.
